### ISSUE TITLE
v1.17.0 - vsphere_virtual_machine : Invalid operation for device '0'

### ISSUE THREAD

**Issue by derrar05aiss** [hashicorp/terraform-provider-vsphere#1016]
Hello,

I'm currently unable to deploy VMs using the terraform-provider-vsphere 1.17.0 and I'm getting : 

vsphere_virtual_machine.svvm: Creating...
vsphere_virtual_machine.svvm: Still creating... [10s elapsed]
Error: error reconfiguring virtual machine: error reconfiguring virtual machine: Invalid operation for device '0'.

with the version 1.15, the VM is created with the same configuration

### Terraform Version

Terraform v0.12.24

### vSphere Provider Version

provider.vsphere = 1.17.0

### Affected Resource(s)

- `vsphere_virtual_machine`

### Terraform Configuration Files

```
variable "dns_servers" {}
variable "domain" {}
variable "vsphere_datacenter" {}
variable "vsphere_cluster" {}
variable "vsphere_datastore" {}
variable "vsphere_template" {}
variable "vsphere_folder" {}
variable "vsphere_network" {}

data "vsphere_datacenter" "datacenter" {
        name = var.vsphere_datacenter
}

data "vsphere_resource_pool" "cluster" {
        name = "${var.vsphere_cluster}/Resources"
        datacenter_id = data.vsphere_datacenter.datacenter.id
}

data "vsphere_datastore" "datastore" {
        name = var.vsphere_datastore
        datacenter_id = data.vsphere_datacenter.datacenter.id
}

data "vsphere_virtual_machine" "template" {
        name          = var.vsphere_template
        datacenter_id = data.vsphere_datacenter.datacenter.id
}

data "vsphere_network" "network" {
        name = var.vsphere_network
        datacenter_id = data.vsphere_datacenter.datacenter.id
}

resource "vsphere_virtual_machine" "svvm" {
        name             = "testvm5"
        folder           = ""
        num_cpus         = 1
        memory           = 1024

        memory_hot_add_enabled = true
        cpu_hot_add_enabled    = true
        cpu_hot_remove_enabled = true

        datastore_id     = data.vsphere_datastore.datastore.id
        resource_pool_id = data.vsphere_resource_pool.cluster.id
        guest_id         = data.vsphere_virtual_machine.template.guest_id
        scsi_type        = data.vsphere_virtual_machine.template.scsi_type

        network_interface {
                network_id     = data.vsphere_network.network.id
        }

        disk {
                label            = "disk0"
                size             = data.vsphere_virtual_machine.template.disks.0.size
                eagerly_scrub    = data.vsphere_virtual_machine.template.disks.0.eagerly_scrub
                thin_provisioned = data.vsphere_virtual_machine.template.disks.0.thin_provisioned
        }
        clone {
                template_uuid = data.vsphere_virtual_machine.template.id
                customize {
                        timeout = 30
                        linux_options {
                          host_name = "testvm"
                          domain = "vm.lab.platform-essential.com"
                        }
                }
        }
}
```

### Debug Output

https://gist.github.com/derrar05aiss/f29196d83eb90dc21380b11eca8c39d5

### Expected Behavior

VM creation

### Actual Behavior

Error while reconfiguring the VM

### Steps to Reproduce

Please list the steps required to reproduce the issue, for example:
1. `terraform apply`

### References

Are there any other GitHub issues (open or closed) or Pull Requests that should
be linked here? 
- https://github.com/terraform-providers/terraform-provider-vsphere/issues/965

**Comment by eystsen** [comment]
Hi,

I'm getting this error and been trying to find a workaround until the issue is fixed. 
Hopefully this observation might help solving it.

**tl;dr;** 
Using terraform with an ubuntu 18.04 created using packer results in the error.
Create the same VM manually and it works fine with terraform.

**More info**
I'm using Ubuntu 18.04 and 16.04 (both server). Due to netplan they behave a bit differently. But the error is exactly the same.  Invalid operation for device '0'.

**Reproducing error**
1. Create an Ubuntu 18.04 VM with packer (distributed switch port group)
2. This VM works fine by itself. Network also works when booting it.
3. Using this VM as a template in terraform results network being disconnected due to "Invalid operation for device '0'.
4. Any changes to the template does not help (ex. removing network adapter)

**Workaround**
1. Create the same Ubuntu 18.04 VM manually. Installing it through vcenter.
(of course something is different between these VMs, but the point is that a minimal installed VM works)
2. Using this VM as a template with terraform works fine

**Summary**
It's not purely a terraform issue, since terraform works with a VM created manually. 
It's also not purely a packer issue, since the VM created by packer works by itself.
The issue can probably be fixed by both packer and terraform.

**Versions tested**
Packer v1.5.4 and v1.5.5
Terraform v0.12.23 and v0.12.24
  vSphere provider v1.17.0

**Comment by vfiftyfive** [comment]
I've got exactly the same issue from packer created VM. Seems to be ok with VM created manually.
Is anybody looking at this?

Nic

**Comment by FrederikNS** [comment]
We just hit this one on 1.17.3, it looks identical to #965.

Terraform version was 0.12.24

**Comment by LennertMertens** [comment]
Having this issue with:
```
Terraform v0.12.24
+ provider.vsphere v1.20.0
```

I'm wondering if this is might be a Packer or Terraform related issue?

**Comment by eystsen** [comment]
> I'm wondering if this is might be a Packer or Terraform related issue?

It could even be both. 
Packer does something that it shouldn't, leaving the VM in an odd state
Terraform should handle that and make it work anyhow

**Comment by rmacian** [comment]
Same here, I have two Vsphere environments and the error only happens in one of them:
**Works on**
Vsphere 6.7.0.42000: 
provider.vsphere v1.40.0
Packer: 1.6.1
Terraform v0.12.26

**Does NOT work**
Vsphere 6.7.0.44000:
BUT:
- I can create a VM manually from vsphere with the template generated with packer and runs ok
- I can create a VM from terraform with an image not generated with Packer

**Comment by arothste-blk** [comment]
I'm having the same problem with
vSphere 7.0.0.10600
packer 1.6.1
terraform 0.13.0 with the v1.23 of the vsphere provider

packer packed VM templates instantiate through the vSphere UI without problem
instantiation with Terraform yield
"error reconfiguring virtual machine: Invalid operation for device '0'"

**Comment by rmacian** [comment]
Definetively the issue is in the provider as pointed by @derrar05aiss

I can create the VM forcing the provider to 1.15

**Comment by martijnxd** [comment]
Yes same issue here : error reconfiguring virtual machine: error reconfiguring virtual machine: Invalid operation for device '0'.
vsphere 7 and hashicorp/vsphere v1.24.3...

**Comment by sestegra** [comment]
I found the root cause of this issue by investigating differences between 1.15.0 and 1.16.0 releases.

### Investigation logs
First, i activated `client_debug` in `vsphere` provider to log SOAP requests and responses into `~/.govmomi/` folder.
A difference appears in `ReconfigVM_Task` request. In fact, an additional `edit` operation on hard drive is present in the 1.16.0 request. This operation appears at rank `0`.
```
<deviceChange xmlns:XMLSchema-instance="http://www.w3.org/2001/XMLSchema-instance" XMLSchema-instance:type="VirtualDeviceConfigSpec">
    <operation>edit</operation>
    <device XMLSchema-instance:type="VirtualDisk">
        <key>2000</key>
        <deviceInfo XMLSchema-instance:type="Description">
            <label>Hard disk 1</label>
            <summary>20,480,000 KB</summary>
        </deviceInfo>
```
The related error message is following.
```
<error>
    <fault xsi:type="InvalidDeviceOperation">
        <property>virtualDeviceSpec.device</property>
        <deviceIndex>0</deviceIndex>
        <badOp>edit</badOp>
    </fault>
    <localizedMessage>Invalid operation for device &apos;0&apos;.</localizedMessage>
</error>
```

**So, the first clue seems to be the `disk` device.**

Then, I decided to compile every commits between 1.15.0 and 1.16.0 and redo the same test with the compiled plugin by using `dev_overrides` in `~/.terraformrc`. I discovered that the commit 05cd12c4b7dfb34f565960ffec0398ad4b19dbb5 reveals the issue. This commit adds the [Storage Policy Based Management](https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.storage.doc/GUID-720298C6-ED3A-4E80-87E8-076FFF02655A.html)).

**The second clue is the `storage_policy_id` parameter that has been added in 1.16.0 version.**

### Solution
In order to get the name of the Storage Policy at VM and disk levels, follow these instructions:

1. Open your inventory and select your template
2. Select `Configure` and `Policies` to get the names

See below the updates required to fix the issue.
```
data "vsphere_storage_policy" "vm_storage_policy" {
  name = <VM storage policy name>
}

data "vsphere_storage_policy" "disk_storage_policy" {
  name = <Disk storage policy name>
}

resource "vsphere_virtual_machine" "vm" {
  name = "vm"
  ...
  storage_policy_id = data.vsphere_storage_policy.vm_storage_policy.id

  disk {
    ...
    storage_policy_id = data.vsphere_storage_policy.disk_storage_policy.id
  }
}
```

**Comment by gitnetofr** [comment]
Hi, I have the same issue.

### Environment

- vSphere Client version 6.7.0.45000
- Terraform v0.13.3
- Terraform provider vsphere v1.24.3

It crashes when disk size is calculated basing on the template data: 

```
disk {
       size             = data.vsphere_virtual_machine.vsphere_template.disks.0.size
       ...
       }

```

I got `Error: error reconfiguring virtual machine: error reconfiguring virtual machine: Invalid operation for device '0'`

However, if I precise the `size > data than vsphere_template.disks.0.size`, it works well: 

```
disk {
       size             = data.vsphere_virtual_machine.vsphere_template.disks.0.size+1
       ...
       }

```

`Apply complete! Resources: 1 added, 0 changed, 0 destroyed.`

When I create the VM manually I do not need to expand the disk size, the VM is created without error. 

Any idea? 

Thanks a lot. 
;-)

**Comment by isometry** [comment]
@gitnetofr : the root issue appears to result from the `vsphere_virtual_machine` data source and resource types measuring disk size in integer GiB, whilst real-world templates may have non-integer GiB disk sizes (e.g. [Packer/vsphere-iso still expects sizes in integer MiB](https://www.packer.io/docs/builders/vmware/vsphere-iso#storage-configuration)). Integer arithmetic in `terraform-provider-vsphere` results in these non-integer GiB disk sizes being rounded down to the nearest GiB. Unfortunately, when you clone using the rounded down size, you're implicitly telling vSphere to shrink the disk, which is unsupported, and results in the `Invalid operation for device '0'` error.

I've opened #1397 with a fix.

**Comment by isometry** [comment]
I believe this issue was fixed by #1397.

**Comment by tenthirtyam** [comment]
@iBrandyJackson and @appilon can you review if this one can be closed per the fix in the PR mentioned above?

Ryan

**Comment by appilon** [comment]
Will close, if the issue persists after upgrading to v2.0.0 of the provider please feel free to comment for re-evaluation

**Comment by demon-zombie** [comment]
I'm seeing this error on v2.0.2 - even when increasing the size of the vm disk size.

Running latest vsphere v7.0.3c

Template vm is created with Ubuntu cloud images:
* https://cloud-images.ubuntu.com/hirsute/current/hirsute-server-cloudimg-amd64.ova
* https://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-amd64.ova

```
>terraform -v
Terraform v1.1.4
on windows_386
+ provider registry.terraform.io/hashicorp/template v2.2.0
+ provider registry.terraform.io/hashicorp/vsphere v2.0.2
```

```
# Generated by Terragrunt. Sig: nIlQXj57tbuaRZEa
variable "hostname" { default = "TEST" }
variable "folder" { default = "Testing" }
variable "disk_size" { default = "80" }
variable "cpu" { default = "1" }
variable "ram" { default = "1024" }

data "template_file" "user_data" {
  template = file("cloud-init.cfg")
  vars = {
    hostname = var.hostname
  }
}

resource "vsphere_virtual_machine" "vm" {
  name             = var.hostname
  folder           = var.folder

  resource_pool_id = data.vsphere_compute_cluster.ivy.resource_pool_id
  datastore_id     = data.vsphere_datastore.vol01-default.id
  wait_for_guest_net_timeout = 0
  wait_for_guest_ip_timeout  = 0
  
  num_cpus = var.cpu
  memory   = var.ram
  guest_id = data.vsphere_virtual_machine.template.guest_id
  scsi_type = data.vsphere_virtual_machine.template.scsi_type
  
  network_interface {
    network_id   = data.vsphere_network.VLAN275B.id
    adapter_type = data.vsphere_virtual_machine.template.network_interface_types[0]
  }

  disk {
    label            = "disk0"
    size             = var.disk_size
    eagerly_scrub    = data.vsphere_virtual_machine.template.disks.0.eagerly_scrub
    thin_provisioned = data.vsphere_virtual_machine.template.disks.0.thin_provisioned
  }
  
  cdrom {
    client_device = true
  }
  
  vapp {
    properties = {
      hostname = var.hostname
      public-keys = [redacted]
      user-data = base64encode(data.template_file.user_data.rendered)
    }
  }
  
  clone {
    template_uuid = data.vsphere_virtual_machine.template.id
  }
}
```

Output response

```
Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # vsphere_virtual_machine.vm will be created
  + resource "vsphere_virtual_machine" "vm" {
      + boot_retry_delay                        = 10000
      + change_version                          = (known after apply)
      + cpu_limit                               = -1
      + cpu_share_count                         = (known after apply)
      + cpu_share_level                         = "normal"
      + datastore_id                            = "datastore-15012"
      + default_ip_address                      = (known after apply)
      + ept_rvi_mode                            = "automatic"
      + firmware                                = "bios"
      + folder                                  = "Testing"
      + force_power_off                         = true
      + guest_id                                = "ubuntu64Guest"
      + guest_ip_addresses                      = (known after apply)
      + hardware_version                        = (known after apply)
      + host_system_id                          = (known after apply)
      + hv_mode                                 = "hvAuto"
      + id                                      = (known after apply)
      + ide_controller_count                    = 2
      + imported                                = (known after apply)
      + latency_sensitivity                     = "normal"
      + memory                                  = 4096
      + memory_limit                            = -1
      + memory_share_count                      = (known after apply)
      + memory_share_level                      = "normal"
      + migrate_wait_timeout                    = 30
      + moid                                    = (known after apply)
      + name                                    = "UBTS-Test01"
      + num_cores_per_socket                    = 1
      + num_cpus                                = 4
      + poweron_timeout                         = 300
      + reboot_required                         = (known after apply)
      + resource_pool_id                        = "resgroup-14027"
      + run_tools_scripts_after_power_on        = true
      + run_tools_scripts_after_resume          = true
      + run_tools_scripts_before_guest_shutdown = true
      + run_tools_scripts_before_guest_standby  = true
      + sata_controller_count                   = 0
      + scsi_bus_sharing                        = "noSharing"
      + scsi_controller_count                   = 1
      + scsi_type                               = "pvscsi"
      + shutdown_wait_timeout                   = 3
      + storage_policy_id                       = (known after apply)
      + swap_placement_policy                   = "inherit"
      + uuid                                    = (known after apply)
      + vapp_transport                          = [
          + "iso",
        ]
      + vmware_tools_status                     = (known after apply)
      + vmx_path                                = (known after apply)
      + wait_for_guest_ip_timeout               = 0
      + wait_for_guest_net_routable             = true
      + wait_for_guest_net_timeout              = 0

      + cdrom {
          + client_device  = true
          + device_address = (known after apply)
          + key            = (known after apply)
        }

      + clone {
          + template_uuid = "4225acde-9766-e25c-3156-0ad3c4f4db9d"
          + timeout       = 30
        }

      + disk {
          + attach            = false
          + controller_type   = "scsi"
          + datastore_id      = "<computed>"
          + device_address    = (known after apply)
          + disk_mode         = "persistent"
          + disk_sharing      = "sharingNone"
          + eagerly_scrub     = false
          + io_limit          = -1
          + io_reservation    = 0
          + io_share_count    = 0
          + io_share_level    = "normal"
          + keep_on_remove    = false
          + key               = 0
          + label             = "disk0"
          + path              = (known after apply)
          + size              = 80
          + storage_policy_id = (known after apply)
          + thin_provisioned  = true
          + unit_number       = 0
          + uuid              = (known after apply)
          + write_through     = false
        }

      + network_interface {
          + adapter_type          = "vmxnet3"
          + bandwidth_limit       = -1
          + bandwidth_reservation = 0
          + bandwidth_share_count = (known after apply)
          + bandwidth_share_level = "normal"
          + device_address        = (known after apply)
          + key                   = (known after apply)
          + mac_address           = (known after apply)
          + network_id            = "dvportgroup-6013"
        }

      + vapp {
          + properties = {
              + "hostname"    = "UBTS-Test01"
			  [redacted]
            }
        }
    }
	
Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

vsphere_virtual_machine.vm: Creating...
vsphere_virtual_machine.vm: Still creating... [10s elapsed]
╷
│ Error: error reconfiguring virtual machine: error reconfiguring virtual machine: Invalid operation for device '0'.
│
│   with vsphere_virtual_machine.vm,
│   on vm.tf line 15, in resource "vsphere_virtual_machine" "vm":
│   15: resource "vsphere_virtual_machine" "vm" {
│
╵
time=2022-01-30T04:56:18-08:00 level=error msg=1 error occurred:
        * exit status 1
```
